File: CPP_Solutions/Matrix_Chain_Multiplication/Palindrome_Partition/PP_Recursive.cpp
#include <bits/stdc++.h>
#include<algorithm>
using namespace std;

int PP_Recursive(string s, int i, int j){
    string s_reversed = s;
    reverse(s_reversed.begin(), s_reversed.end()); // Reveresed String Obtained
    if(i>=j)
        return 0;
    if(s == s_reversed) // Check if original string is a boolean
        return  0;
    int min = INT_MAX, temp;
    for(int k = i; k < j; k++){
        temp = PP_Recursive(s, i, k) + PP_Recursive(s, k + 1, j) + 1; // We add this one since original string has 
        // already been partitioned to solve the subproblems. So that initial partition counts as 1
        if(temp<min)
            min = temp;
    }
    return min;
}

int main()
{
    string s = "geek";
    int n = s.length();
    cout << PP_Recursive(s, 1, n-1);

    return 0;
}

File: Python_Solutions/Matrix_Chain_Multiplication/Palindrome_Partition/PP_Recursive.py
import sys

def PP_Recursive(s,i,j):
    rev_s = s[::-1]
    if(i >= j):
        return 0
    if(s == rev_s): # Checking if original string is boolean
        return 0
    minPartition = sys.maxsize
    for k in range(i,j):
        temp = PP_Recursive(s, i, k) + PP_Recursive(s, k+1, j) + 1 # We add this one since original string has already
        # been partitioned to solve the subproblems. So that initial partition counts as 1
        if(temp < minPartition):
            minPartition = temp
    return minPartition

s = "geek"
n = len(s)
print(PP_Recursive(s, 1, n - 1))